eth/handler_eth.go

```diff
@@ -148,6 +152,26 @@ func (h *ethHandler) handleHeaders(peer *eth.Peer, headers []*types.Header) erro
 			}
 			peer.Log().Debug("Whitelist block verified", "number", headers[0].Number.Uint64(), "hash", want)
 		}
+
+		// fetch checkpoint from heimdall for whitelist
+		if headers[0].Number.Uint64()%h.chain.Config().Bor.Sprint == 0 && !h.chain.Engine().(*bor.Bor).WithoutHeimdall {
+			endBlockNum, endBlockHash, ok := h.whitelistCheckpoint()
```

Comment:
You are fetching checkpoint every time a block sync. That will slow down everything and won't work. You have to keep getting new checkpoints separately and updating the `whitelistCheckpoint`. So block sync just checks the whitelist checkpoint instead of fetching everytime.